docs/cplusplus-style-guide.md

```diff
@@ -0,0 +1,585 @@
+Please note, the *default* delete policy for these types uses the **default** C++ delete operator for **single** objects. However, the API of these types offers custom deleter and allocator functions for managing non-default allocation/deallocation policies such as needed when managing **array** object types.
+
+Below are some types which guarantee to manage **array** object types (continuous block of memory) by default, and other types which do memory management as a side effect of their primary business logic.
+
+* `std::vector<...>` guarantees to manage a continuous block of memory (array) by definition.
+* `std::string<...>` guarantees to manage a continuous block of memory (array) by definition.
+* `SharedArray<...>` and `Array<...>` classes (from `fetch::memory` namespace) that guarantee to manage a continuous block of memory (array) by definition.
+* Classes from `fetch::byte_array` namespace, such as `BasicByteArray` 
+
+### Transfer between ownership models
+
+It is safe to transfer ownership from `std::unique_ptr` to `std::shared_ptr`.
+
+However, do **NOT** transfer ownership the other way around, i.e. from `std::shared_ptr` to `std::unique_ptr`.
+
+Any justification for doing this is highly questionable. If you really feel it is necessary to do this, then keep the following in mind:
+   
+* It is a **non**-trivial operation.
+* The `shared_ptr` type is **NOT** designed for this.
+* If you still want to proceed, then read <a href="https://stackoverflow.com/questions/15337461/move-ownership-from-stdshared-ptr-to-stdunique-ptr" target=_blank>this</a> which may provide some guidance.
+
+
+Example of using smart pointer with **shared** ownership:
+
+``` c++
+#include <memory>
+#include <iostream>
+
+//* Parent Scope block
+{
+  std::shared_ptr<int> a0;
+  
+  //* Nested scope block
+  {
+    std::shared_ptr<int> a1 = std::make_shared<int>(5); //* refcount=1
+    a0 = a1; //* The `a1` instance become shared at this point (refcount=2)
+    
+    //* BOTH instances are valid, controlling the same `int` instance:
+    std::cout << "a0 = " << *a0 << ", a1 = " << *a1 << std::endl;
+    
+    throw std::exception();
+    
+    //* The exception will cause stack unwinding, during which 
+    //* is the destructor of `a1` instance called automatically, 
+    //* what decrements refcount of shared object (to 1).
+    //* At this point, the `a0` shared object is controlling lifecycle
+    //* of the `int` instance created above in this scope.
+  }
+  
+  //* We are still in stack unwinding process here (due to earlier exception),
+  //* during which automatically called destructor of `a0` instance  decrements
+  //* refcount of shared object (to 0), what finally results to destruction of
+  //* the `int` instance allocated & initialised in the nested scope above.
+}
+```
+
+
+Example of using smart pointer with **exclusive** ownership:
+``` c++
+#include <memory>
+
+//* Parent Scope block
+{
+  std::unique_ptr<int> a0;
+  
+  //* Nested scope block
```

Comment:
And again here - also not sure if we want to include this section as it basically an copy and paste from the one above